sdk/python/kfp/compiler_cli_tests/test_data/lightweight_python_functions_v2_pipeline.yaml

```diff
@@ -101,40 +105,36 @@ deploymentSpec:
           \ Model artifact, which has a .metadata dictionary\n    # to store arbitrary\
           \ metadata for the output artifact.\n    model.metadata['accuracy'] = 0.9\n\
           \n"
-        args:
-        - --executor_input
-        - '{{$}}'
-        - --function_to_execute
-        - train
+        image: python:3.7
 components:
   comp-preprocess:
     inputDefinitions:
       parameters:
-        message:
-          parameterType: STRING
         input_dict_parameter:
           parameterType: STRUCT
+        message:
```

Comment:
I'm curious why do we see such diffs? Does that mean the order is nondeterministic? Or the order depends on the specific version of a protobuf library?

/cc @connor-mccarthy

Reply:
It appears only `dsl.Condition` and `dsl.ParallelFor` compilation steps produce non-deterministically ordered YAML.

I feel reasonably confident that this is not attributed to the actual "write yaml" code, but rather to how these conditions/loops are represented in the in-memory IR. I see we're using a `defaultdict(set)` [here](https://github.com/kubeflow/pipelines/blob/c7301c449c3c857e4f9dba09af4911f1a3abf8f2/sdk/python/kfp/compiler/compiler.py#L501) to store this data and I think the use of `set` could explain the non-determinism.

This probably warrants another PR to investigate: swapping out the `defaultdict(set)` for `defaultdict(list)` (and updating downstream methods from `.add` to `.append`) to see if this resolves the issue.

Reply:
I think we might need to bring back `sort_keys=True` when exporting json/yaml. Not blocking this PR though.